### PR TITLE
Fix for activerecord join dependency instantiate bug

### LINKED ISSUE #19896: Child records not properly loaded when accessible through multiple association paths
If you have a model that can reference another through multiple association paths, and you load both paths with `.includes`, and you use `.references` in your finder, some references to the "other" model's children don't work properly.

```ruby
unless File.exist?('Gemfile')
  File.write('Gemfile', <<-GEMFILE)
    source 'https://rubygems.org'
    gem 'rails', github: 'rails/rails'
    gem 'arel', github: 'rails/arel'
    gem 'rack', github: 'rack/rack'
    gem 'i18n', github: 'svenfuchs/i18n'
    gem 'sqlite3'
  GEMFILE

  system 'bundle'
end

require 'bundler'
Bundler.setup(:default)

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :customers
  create_table :phone_numbers do |t|
    t.belongs_to :customer
  end
  create_table :orders do |t|
    t.belongs_to :customer
  end
  create_table :receipts do |t|
    t.belongs_to :order
    t.belongs_to :customer
  end
end

class PhoneNumber < ActiveRecord::Base
  belongs_to :customer
end

class Customer < ActiveRecord::Base
  has_many :orders
  has_many :phone_numbers
end

class Order < ActiveRecord::Base
  belongs_to :customer
  has_many :receipts
end

class Receipt < ActiveRecord::Base
  belongs_to :order
  belongs_to :customer
end

class Test < ActiveSupport::TestCase

  def setup
    customer = Customer.new
    customer.phone_numbers.build
    customer.save!
    order = Order.create(customer: customer)
    Receipt.create(order: order, customer: customer)
  end

  test 'through receipt.customer, with .references and .includes' do
    receipt = Receipt.references(:orders).includes(order: {customer: :phone_numbers}, customer: :phone_numbers).first
    assert_equal 1, receipt.customer.phone_numbers.count
    assert_equal 1, receipt.customer.phone_numbers.to_a.length # This fails
  end

  test 'through receipt.customer, no .references' do
    receipt = Receipt.includes(order: {customer: :phone_numbers}, customer: :phone_numbers).first
    assert_equal 1, receipt.customer.phone_numbers.count
    assert_equal 1, receipt.customer.phone_numbers.to_a.length
  end

  test 'through receipt.customer, no .includes' do
    receipt = Receipt.references(:orders).first
    assert_equal 1, receipt.customer.phone_numbers.count
    assert_equal 1, receipt.customer.phone_numbers.to_a.length
  end

  test 'through receipt.order.customer, with .references and .includes' do
    receipt = Receipt.references(:orders).includes(order: {customer: :phone_numbers}, customer: :phone_numbers).first
    assert_equal 1, receipt.order.customer.phone_numbers.count
    assert_equal 1, receipt.order.customer.phone_numbers.to_a.length
  end

end
```

### PR DESCRIPTION
**Pull request by meinac**
This fix closes #19896

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Could you add a test case that assets the new behavior?

**Comment by meinac** [comment]
@rafaelfranca I've created a test case for testing the behaviour. By the way, data model of the original issue is wrong to me but even if the data model is wrong rails should handle relations correctly so this patch does this.

File: activerecord/test/models/mentor.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Mentor < ActiveRecord::Base
+
```
Comment (rafaelfranca) [inline activerecord/test/models/mentor.rb]: :scissors:

File: activerecord/test/models/mentor.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+  has_many :developers
+
```
Comment (rafaelfranca) [inline activerecord/test/models/mentor.rb]: :scissors:

File: activerecord/test/cases/associations/eager_test.rb [hunk 3]
```diff
@@ -1214,6 +1216,16 @@ def test_deep_including_through_habtm
     assert_no_queries { assert_equal 2, posts[1].categories[0].categorizations.length }
   end
 
+  def test_eager_load_multiple_associations_with_references
+    mentor = Mentor.create!(name: "Barış Can DAYLIK")
+    developer = Developer.create!(name: "Mehmet Emin İNAÇ", mentor: mentor)
+    Contract.create!(developer: developer)
+    project = Project.create!(name: "VNGRS", mentor: mentor)
+    project.developers << developer
+    projects = Project.references(:mentors).includes(mentor: {developers: :contracts}, developers: :contracts)
```
Comment (rafaelfranca) [inline activerecord/test/cases/associations/eager_test.rb]: space after `{` and before `}`

**Comment by meinac** [comment]
@rafaelfranca thanks for you response, I've fixed minor problems.

**Comment by meinac** [comment]
@rafaelfranca did you have a chance to look into this?

**Pull request merged**